File: backend/api/v1/repair.go
package v1

import (
	"net/http"
	"strconv"
	"time"

	"github.com/ems/backend/internal/dto"
	"github.com/ems/backend/internal/middleware"
	"github.com/ems/backend/internal/model"
	"github.com/ems/backend/internal/service"
	"github.com/gin-gonic/gin"
)

var (
	repairOrderService *service.RepairOrderService
)

func InitRepair() {
	repairOrderService = service.NewRepairOrderService()
}

// =====================================================
// Repair Order APIs
// =====================================================

// CreateRepairOrder creates a new repair order
// @Summary Create repair order
// @Tags repair
// @Router /repair/orders [post]
func CreateRepairOrder(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	var req dto.CreateRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	createReq := &service.CreateOrderRequest{
		EquipmentID:      req.EquipmentID,
		FaultDescription: req.FaultDescription,
		FaultCode:        req.FaultCode,
		Photos:           req.Photos,
		Priority:         req.Priority,
		ReporterID:       userID,
	}

	order, err := repairOrderService.CreateOrder(createReq)
	if err != nil {
		handleServiceError(c, err)
		return
	}

	c.JSON(http.StatusCreated, orderToResponse(order))
}

// ListRepairOrders returns repair orders
// @Summary Get repair orders
// @Tags repair
// @Router /repair/orders [get]
func ListRepairOrders(c *gin.Context) {
	var query dto.RepairTaskQuery
	if err := c.ShouldBindQuery(&query); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Set default pagination
	if query.Page == 0 {
		query.Page = 1
	}
	if query.PageSize == 0 {
		query.PageSize = 20
	}

	filter := &service.RepairOrderFilter{
		Status:     query.Status,
		Priority:   query.Priority,
		AssignedTo: query.AssignedTo,
		Page:       query.Page,
		PageSize:   query.PageSize,
	}

	if query.DateFrom != "" {
		if t, err := time.Parse("2006-01-02", query.DateFrom); err == nil {
			filter.DateFrom = t
		}
	}
	if query.DateTo != "" {
		if t, err := time.Parse("2006-01-02", query.DateTo); err == nil {
			filter.DateTo = t
		}
	}

	result, err := repairOrderService.List(filter)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	response := make([]dto.RepairOrderResponse, len(result.Items))
	for i, order := range result.Items {
		response[i] = orderToResponse(&order)
	}

	c.JSON(http.StatusOK, dto.RepairOrderListResponse{
		Total: result.Total,
		Items: response,
	})
}

// GetRepairOrder returns a single repair order
// @Summary Get repair order
// @Tags repair
// @Router /repair/orders/:id [get]
func GetRepairOrder(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	order, err := repairOrderService.GetByID(uint(id))
	if err != nil {
		if err == service.ErrNotFound {
			c.JSON(http.StatusNotFound, gin.H{"error": "Order not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, dto.RepairOrderDetailResponse{
		RepairOrderResponse: orderToResponse(order),
	})
}

// AssignRepairOrder assigns a repair order to a technician
// @Summary Assign repair order
// @Tags repair
// @Router /repair/orders/:id/assign [post]
func AssignRepairOrder(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var req dto.AssignRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	order, err := repairOrderService.AssignOrder(uint(id), req.AssignTo, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, orderToResponse(order))
}

// StartRepair starts a repair task
// @Summary Start repair
// @Tags repair
// @Router /repair/orders/:id/start [post]
func StartRepair(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var req dto.StartRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := repairOrderService.StartRepair(uint(id), userID); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Repair started"})
}

// UpdateRepair updates repair progress
// @Summary Update repair progress
// @Tags repair
// @Router /repair/orders/:id/update [post]
func UpdateRepair(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var req dto.UpdateRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updateReq := &service.UpdateRepairRequest{
		Solution:    req.Solution,
		SpareParts:  req.SpareParts,
		ActualHours: req.ActualHours,
		Photos:      req.Photos,
		NextStatus:  req.NextStatus,
	}

	if err := repairOrderService.UpdateRepair(uint(id), userID, updateReq); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Repair updated"})
}

// ConfirmRepair confirms or rejects a completed repair
// @Summary Confirm repair
// @Tags repair
// @Router /repair/orders/:id/confirm [post]
func ConfirmRepair(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var req dto.ConfirmRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := repairOrderService.ConfirmRepair(uint(id), userID, req.Accepted, req.Comment, req.Photos); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Repair confirmed"})
}

// AuditRepair audits a confirmed repair
// @Summary Audit repair
// @Tags repair
// @Router /repair/orders/:id/audit [post]
func AuditRepair(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	// Check permissions
	if role, _ := middleware.GetUserRole(c); role != "admin" && role != "engineer" && role != "supervisor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "Insufficient permissions"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var req dto.AuditRepairRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := repairOrderService.AuditRepair(uint(id), userID, req.Approved, req.Comment, &req.ActualHours); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Repair audited"})
}

// GetMyRepairTasks returns current user's active repair tasks
// @Summary Get my repair tasks
// @Tags repair
// @Router /repair/my-tasks [get]
func GetMyRepairTasks(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	tasks, err := repairOrderService.GetMyTasks(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	response := make([]dto.RepairOrderResponse, len(tasks))
	for i, task := range tasks {
		response[i] = orderToResponse(&task)
	}

	c.JSON(http.StatusOK, response)
}

// GetMyRepairStatistics returns current user's repair statistics
// @Summary Get my repair statistics
// @Tags repair
// @Router /repair/my-stats [get]
func GetMyRepairStatistics(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	stats, err := repairOrderService.GetUserStatistics(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, dto.MyRepairStatistics{
		PendingCount:    stats.PendingCount,
		InProgressCount: stats.InProgressCount,
		CompletedCount:  stats.CompletedCount,
		TodayCompleted:  stats.TodayCompleted,
	})
}

// GetRepairStatistics returns repair statistics
// @Summary Get repair statistics
// @Tags repair
// @Router /repair/statistics [get]
func GetRepairStatistics(c *gin.Context) {
	stats, err := repairOrderService.GetStatistics()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, dto.RepairStatistics{
		TotalOrders:      stats.TotalOrders,
		PendingOrders:    stats.PendingOrders,
		InProgressOrders: stats.InProgressOrders,
		CompletedOrders:  stats.CompletedOrders,
		TodayCompleted:   stats.TodayCompleted,
		TodayCreated:     stats.TodayCreated,
		AvgRepairTime:    stats.AvgRepairTime,
		AvgResponseTime:  stats.AvgResponseTime,
	})
}

// =====================================================
// Helper Functions
// =====================================================

func orderToResponse(order *model.RepairOrder) dto.RepairOrderResponse {
	r := dto.RepairOrderResponse{
		ID:               order.ID,
		EquipmentID:      order.EquipmentID,
		FaultDescription: order.FaultDescription,
		FaultCode:        order.FaultCode,
		ReporterID:       order.ReporterID,
		Status:           dto.RepairStatus(order.Status),
		Priority:         order.Priority,
		Photos:           order.Photos,
		Solution:         order.Solution,
		CreatedAt:        order.CreatedAt,
		StartedAt:        order.StartedAt,
		CompletedAt:      order.CompletedAt,
		ConfirmedAt:      order.ConfirmedAt,
		AuditedAt:        order.AuditedAt,
	}

	if order.Equipment.ID > 0 {
		r.EquipmentCode = order.Equipment.Code
		r.EquipmentName = order.Equipment.Name
	}
	if order.Reporter.ID > 0 {
		r.ReporterName = order.Reporter.Name
	}
	if order.Assignee != nil && order.Assignee.ID > 0 {
		r.AssignedTo = order.Assignee.ID
		r.AssigneeName = order.Assignee.Name
	}

	return r
}

File: frontend/src/api/repair.ts
import request from './request'

// Types
export type RepairStatus = 'pending' | 'assigned' | 'in_progress' | 'testing' | 'confirmed' | 'audited' | 'closed'

export interface RepairOrder {
  id: number
  equipment_id: number
  equipment_code?: string
  equipment_name?: string
  fault_description: string
  fault_code?: string
  reporter_id: number
  reporter_name?: string
  assigned_to?: number
  assignee_name?: string
  status: RepairStatus
  priority: number // 1=高,2=中,3=低
  photos?: string[]
  solution?: string
  spare_parts?: string
  actual_hours?: number
  created_at: string
  started_at?: string
  completed_at?: string
  confirmed_at?: string
  audited_at?: string
}

export interface RepairOrderListResponse {
  total: number
  items: RepairOrder[]
}

export interface RepairOrderDetail extends RepairOrder {
  logs?: RepairLog[]
}

export interface RepairLog {
  id: number
  order_id: number
  user_id: number
  user_name?: string
  action: string
  content: string
  created_at: string
}

export interface CreateRepairRequest {
  equipment_id: number
  fault_description: string
  fault_code?: string
  photos?: string[]
  priority?: number
}

export interface AssignRepairRequest {
  assign_to: number
}

export interface StartRepairRequest {
  latitude?: number
  longitude?: number
}

export interface UpdateRepairRequest {
  solution?: string
  spare_parts?: string
  actual_hours?: number
  photos?: string[]
  next_status?: string // testing, confirmed
}

export interface ConfirmRepairRequest {
  accepted: boolean
  comment?: string
  photos?: string[]
}

export interface AuditRepairRequest {
  approved: boolean
  comment?: string
  actual_hours?: number
}

export interface RepairStatistics {
  total_orders: number
  pending_orders: number
  in_progress_orders: number
  completed_orders: number
  today_completed: number
  today_created: number
  avg_repair_time: number
  avg_response_time: number
}

export interface MyRepairStatistics {
  pending_count: number
  in_progress_count: number
  completed_count: number
  today_completed: number
}

// Repair Order APIs
export const repairOrderApi = {
  getOrders: (params: {
    page?: number
    page_size?: number
    status?: string
    priority?: number
    assigned_to?: number
    date_from?: string
    date_to?: string
  }) => request.get<RepairOrderListResponse>('/repair/orders', { params }),

  getOrder: (id: number) => request.get<RepairOrderDetail>(`/repair/orders/${id}`),

  createOrder: (data: CreateRepairRequest) =>
    request.post<RepairOrder>('/repair/orders', data),

  assignOrder: (id: number, data: AssignRepairRequest) =>
    request.post<RepairOrder>(`/repair/orders/${id}/assign`, data),

  startRepair: (id: number, data: StartRepairRequest) =>
    request.post(`/repair/orders/${id}/start`, data),

  updateRepair: (id: number, data: UpdateRepairRequest) =>
    request.post(`/repair/orders/${id}/update`, data),

  confirmRepair: (id: number, data: ConfirmRepairRequest) =>
    request.post(`/repair/orders/${id}/confirm`, data),

  auditRepair: (id: number, data: AuditRepairRequest) =>
    request.post(`/repair/orders/${id}/audit`, data),

  getMyTasks: () => request.get<RepairOrder[]>('/repair/my-tasks'),

  getMyStats: () => request.get<MyRepairStatistics>('/repair/my-stats'),

  getStatistics: () => request.get<RepairStatistics>('/repair/statistics'),
}

// Helper functions
export const getRepairStatusText = (status: RepairStatus): string => {
  const map: Record<RepairStatus, string> = {
    pending: '待派单',
    assigned: '已派单',
    in_progress: '维修中',
    testing: '待测试',
    confirmed: '待审核',
    audited: '已审核',
    closed: '已关闭',
  }
  return map[status] || status
}

export const getRepairStatusType = (status: RepairStatus): string => {
  const map: Record<RepairStatus, string> = {
    pending: 'info',
    assigned: 'primary',
    in_progress: 'warning',
    testing: 'warning',
    confirmed: 'success',
    audited: 'success',
    closed: 'info',
  }
  return map[status] || 'info'
}

export const getPriorityText = (priority: number): string => {
  const map: Record<number, string> = {
    1: '高',
    2: '中',
    3: '低',
  }
  return map[priority] || '中'
}

export const getPriorityType = (priority: number): string => {
  const map: Record<number, string> = {
    1: 'danger',
    2: 'warning',
    3: 'info',
  }
  return map[priority] || 'info'
}

// Alias exports for backward compatibility
export const getStatusType = getRepairStatusType
export const getStatusText = getRepairStatusText
